File: CSE216_Assignment4/BinarySearchTree.java

import java.util.*;


public class BinarySearchTree<T extends Comparable<T>> extends Thread implements Iterable<T>{
    private String name;
    private Node<T> ptr;
    private List<T> sorted;
    private int iter;

    public BinarySearchTree() {

    }

    public static void main(String... args) {
        // each tree has a name, provided to its constructor
        BinarySearchTree<Integer> t1 = new BinarySearchTree<>("Oak");

        // adds the elements to t1 in the order 5, 3, 0, and then 9
        t1.addAll(Arrays.asList(5, 3, 0, 8));
        BinarySearchTree<Integer> t2 = new BinarySearchTree<>("Maple");

        // adds the elements to t2 in the order 9, 5, and then 10
        t2.addAll(Arrays.asList(9, 7, 10));

        System.out.println(t1); // see the expected output for exact format
        t1.forEach(System.out::println); // iteration in increasing order

        System.out.println(t2); // see the expected output for exact format
        t2.forEach(System.out::println); // iteration in increasing order

        BinarySearchTree<String> t3 = new BinarySearchTree<>("Cornucopia");
        t3.addAll(Arrays.asList("coconut", "apple", "banana", "plum", "durian", "no durians on this tree!", "tamarind"));

        System.out.println(t3); // see the expected output for exact format
        t3.forEach(System.out::println); // iteration in increasing order

        List<BinarySearchTree<Integer>> bstlist = new ArrayList<>();
        bstlist.add(t1);
        bstlist.add(t2);
        merge(bstlist);
    }

    private void addAll(List<T> asList) {
        for (T t : asList) {
            insert(t);
        }
        List<T> copy;
        copy = asList;

        Collections.sort(copy);
        sorted = copy;
    }


    public void insert(T key){
        ptr = insertRec(ptr, key);
    }

    public Node<T> insertRec(Node<T> ptr, T key){
        if(ptr == null){
            ptr = new Node<>(key);
            return ptr;
        }
        int compare = ptr.val.compareTo(key);
        if (compare > 0)
            ptr.left = insertRec(ptr.left, key);
        else if (compare < 0)
            ptr.right = insertRec(ptr.right, key);

        return ptr;
    }


    public BinarySearchTree(String name){
        this.name = name;
        ptr = null;
        sorted = new ArrayList<>();
        iter = 0;
    }

    @Override
    public Iterator<T> iterator() {
        //use next method to create traversal of ordered list
        Iterator<T> newIterator = new Iterator<T>() {
            @Override
            public boolean hasNext() {
                return iter != sorted.size();
            }

            @Override
            public T next() {
                return sorted.get(iter++);
            }

        };
        return newIterator;
    }


    @Override
    public String toString(){

        return "[" + name + "] " + toStringPreorder(ptr);
    }

    public String toStringPreorder(Node<T> ptr) {
        String s = "";
        if (ptr == null) {
            return "";
        }

        s += ptr.val.toString();
        if(ptr.left != null){
            s += " L: (" + toStringPreorder(ptr.left) + ") ";
        }
        if(ptr.right != null){
            s += " R: (" + toStringPreorder(ptr.right) + ") ";
        }
        return s;
    }

    public static <T extends Comparable<T>> List<T> merge(List<BinarySearchTree<T>> bstlist){
        List<T> bigList = new ArrayList<>();

        List<BinarySearchTree<T>> threadList = new ArrayList<>();

        for(int i = 0; i< bstlist.size(); i++){
            BinarySearchTree<T> thread = new BinarySearchTree<>();
            threadList.add(thread);
            thread.start();
        }
        ArrayList pointers = new ArrayList(bstlist.size());






        return bigList;

    }
    @Override
    public void run(){

    }


    class Node<T>{
        private final T val;
        private Node<T> left;
        private Node<T> right;

        Node(T t){this.val = t;}

        //take small lists(sorted) and merge into big one
        //global list
        //have a thread that deals with one list
        //use extend thread class also get lock of the big list
    }
}

File: CSE216_Assignment4/binarysearchtree.py
class Node:
    right = None
    left = None
    data = None

    def __init__(self, data=None):
        self.right = None
        self.left = None
        self.data = data

    def __str__(self):
        s = str(self.data) + ""
        if self.left is not None:
            s = s + " L:(" + str(self.left) + ")"
        if self.right is not None:
            s = s + " R:(" + str(self.right) + ")"
        return s

    def __iter__(self):
        return inorder(self)


class binarysearchtree:
    name = None
    left = None
    right = None
    root = None

    def __init__(self, name, root):
        self.right = Node()
        self.left = Node()
        self.root = root
        self.name = name

    def add_all(self, *lst):
        self.root = None
        for num in lst:
            self.root = self.insert(self.root, num)

    def insert(self, node, data):
        if node is None:
            return Node(data)
        if data < node.data:
            node.left = self.insert(node.left, data)
        else:
            node.right = self.insert(node.right, data)
        return node

    def __iter__(self):
        return inorder(self)

    def __str__(self):
        return "[" + self.name + "]" + str(self.root)

def inorder(t):
    if t:
        for y in inorder(t.left):
            yield y
        yield t.data

        for y in inorder(t.right):
            yield y


if __name__ == "__main__":
    t1 = binarysearchtree(name="Oak", root=Node())
    t2 = binarysearchtree(name="Birch", root=Node())
    t1.add_all(5, 3, 9, 0)
    t2.add_all(1, 0, 10, 2, 7)
    for x in t1.root:
        print(x)
    for x in t2.root:
        print(x)
    tree = binarysearchtree(name="Oak", root=Node())
    tree.add_all(5, 3, 9, 0)  # adds the elements in the order 5, 3, 9, and then 0
    print(t2)
